src/libs/elektra/kdb.c

```diff
@@ -1324,6 +1381,441 @@ int kdbSet (KDB * handle, KeySet * ks, Key * parentKey)
+ * @param pluginConfig the config KeySet for the plugin; is always consumed, i.e. you shouldn't ksDel() it after calling this
+ * @param errorKey     used for error reporting
+ *
+ * @retval 1 on success
+ * @retval 0 otherwise
+ */
+static int ensurePluginState (KDB * handle ELEKTRA_UNUSED, const char * mountpoint ELEKTRA_UNUSED, const char * pluginName ELEKTRA_UNUSED,
+			      enum PluginContractState pluginState, KeySet * pluginConfig ELEKTRA_UNUSED, Key * errorKey ELEKTRA_UNUSED)
+{
+	switch (pluginState)
+	{
+	case PLUGIN_STATE_UNMOUNTED:
+		ksDel (pluginConfig);
+		return ensurePluginUnmounted (handle, mountpoint, pluginName, errorKey);
+	case PLUGIN_STATE_MOUNTED:
+		return 0; // TODO: ensurePluginMounted (handle, mountpoint, pluginName, pluginConfig, errorKey);
+	case PLUGIN_STATE_REMOUNT:
+		return 0; // TODO: ensurePluginUnmounted (handle, mountpoint, pluginName, errorKey) && ensurePluginMounted (handle,
+			  // mountpoint, pluginName, pluginConfig, errorKey);
+	default:
+		ELEKTRA_ASSERT (0, "missing switch case");
+		return 0;
+	}
+}
+
+/**
+ * This function can be used the given KDB @p handle meets certain conditions,
+ * specified in @p contract. Currently the following conditions are supported:
+ *
+ * - `system/plugins/<mountpoint>/<pluginname>` defines the state of the plugin
+ *   `<pluginname>` for the mountpoint `<mountpoint>`:
+ * 	- The value `unmounted` ensures the plugin is not mounted, at this mountpoint.
+ * 	- The value `mounted` ensures the plugin is mounted, at this mountpoint.
+ * 	  If the plugin is not mounted, we will try to mount it.
+ * 	- The value `remount` always mounts the plugin, at this mountpoint.
+ * 	  If it was already mounted, it will me unmounted and mounted again.
+ * 	  This can be used to ensure the plugin is mounted with a certain configuration.
+ * - Keys below `system/plugins/<mountpoint>/<pluginname>/config` are extracted and used
+ *   as the plugins config KeySet during mounting. `system/plugins/<mountpoint>/<pluginname>`
+ *   will be repleced by `user` in the keynames. If no keys are given, an empty KeySet is used.
+ *
+ * There are a few special values for `<mountpoint>`:
+ * - `global` is used to indicate the plugin should (un)mounted as a global plugin.
+ *   Currently this only supports (un)mounting plugins from/to the subposition `maxonce`.
+ * - `parent` is used to indicate the keyname of @p parentKey shall be used as the mountpoint.
+ *
+ * If `<mountpoint>` is none of those values, it has to be valid keyname with the slashes escaped.
+ * That means it has to start with `/`, `user`, `system`, `dir` or `spec`.
+ *
+ * If `<mountpoint>` is NOT `global`, currently only `unmounted` is supported (not `mounted` and `remounted`).
+ *
+ * NOTE: This function only works properly, if the list plugin is mounted in all global positions.
+ * If this is not the case, 1 will be returned, because this is seen as an implicit condition in the contract.
```

Comment:
Better to fail if you cannot do it. The application has no control over if the list plugin is there or not. And it expects that `kdbEnsure` actually ensured its contract if it returned successfully.

Reply:
The return value 1 indicates a failure. Only 0 means success.

Reply:
This is non-standard compared to other kdb calls. Can you change that to be -1?

Reply:
I could change it to -2, if you want all negative return values + 0 for success...

1 is returned, if the contract could not ensured, but the `kdbEnsure` call was entirely valid. e.g. if a plugin could not be mounted

-1 meanwhile is used to indicate that either the contract is malformed, or something else about the `kdbEnsure` call is wrong.

If 1 is returned, the application should show a meaningful message to the user, because they might need to change their KDB configuration. If -1 is returned, the application probably has a bug, and without changing the code there is little to no chance of fixing the problem...

Reply:
> I could change it to -2, if you want all negative return values + 0 for success...

I do not think we need this distinction in the return value as you can simply set different errors in the errorKey.

Furthermore, I think it might be useful to indicate if something was done at all.
So I would propose following return codes:

- 1: could not run kdbEnsure successfully (error is in the errorKey)
0: kdbEnsure successful but nothing was to do
1: mounted or unmounted something, so the application now has different mountpoints than the rest of the system

This would be basically identical to return codes of the other kdb* calls (and also the resolver).

Iirc there are now also macros defined for these values.